### PR TITLE
Add and use IExecutable and IProcess abstractions

### PR DESCRIPTION
**Pull request by drewnoakes**
Git Extensions uses external processes to obtain data and perform various operations. The primary executable used is `git.exe`, however there are others (`plink`, `gitk`, `pageant`, ...).

There are many places in the code where processes are spawned and their output consumed, making the calling code difficult to test. This is particularly true for `GitModule`, but there are plenty of other places too.

It's difficult to see the extent to which such processes are used, as only a subset of such processes are displayed in the command log.

`System.Diagnostics.Process` also has no direct support for asynchronously consuming process output.

This PR introduces a standard abstraction for the creation and management of processes, and applies it across the code. It adds first class async support, so we no longer need classes such as `SynchronizedProcessReader` that spawn one-off threads just to read ouput streams, reducing memory usage and increasing thread pool utilisation.

## API

The main types are `IExecutable` and `IProcess` (see code for documented versions):

```c#
public interface IExecutable
{
    IProcess Start(
        ArgumentString arguments = default,
        bool createWindow = false,
        bool redirectInput = false,
        bool redirectOutput = false,
        [CanBeNull] Encoding outputEncoding = null);
}

public interface IProcess : IDisposable
{
    StreamWriter StandardInput { get; }
    StreamReader StandardOutput { get; }
    StreamReader StandardError { get; }
    int? ExitCode { get; }
    int? WaitForExit();
    Task<int?> WaitForExitAsync();
    void WaitForInputIdle();
}
```

`IProcess` is quite low level. This PR introduces more ergonomic APIs for the most common operations via stateless extension methods in `ExecutableExtensions`.

```c#
public static class ExecutableExtensions
{
    public static string GetOutput(
        this IExecutable executable,
        ArgumentString arguments = default,
        byte[] input = null,
        Encoding outputEncoding = null,
        CommandCache cache = null,
        bool stripAnsiEscapeCodes = true);

    public static async Task<string> GetOutputAsync(
        this IExecutable executable,
        ArgumentString arguments = default,
        byte[] input = null,
        Encoding outputEncoding = null,
        CommandCache cache = null,
        bool stripAnsiEscapeCodes = true);

    public static bool RunCommand(
        this IExecutable executable,
        ArgumentString arguments = default,
        byte[] input = null,
        bool createWindow = false);

    public static async Task<bool> RunCommandAsync(
        this IExecutable executable,
        ArgumentString arguments = default,
        byte[] input = null,
        bool createWindow = false);

    public static IEnumerable<string> GetOutputLines(
        this IExecutable executable,
        ArgumentString arguments = default,
        byte[] input = null,
        Encoding outputEncoding = null,
        bool stripAnsiEscapeCodes = true);

    public static ExecutionResult Execute(
        this IExecutable executable,
        ArgumentString arguments,
        Action<StreamWriter> writeInput = null,
        Encoding outputEncoding = null,
        bool stripAnsiEscapeCodes = true);

    public static async Task<ExecutionResult> ExecuteAsync(
        this IExecutable executable,
        ArgumentString arguments,
        Action<StreamWriter> writeInput = null,
        Encoding outputEncoding = null,
        bool stripAnsiEscapeCodes = true);
}

public readonly struct ExecutionResult
{
    [CanBeNull] public string StandardOutput { get; }
    [CanBeNull] public string StandardError { get; }
    public int? ExitCode { get; }
    public bool ExitedSuccessfully { get; }
    [NotNull] public string AllOutput { get; }
}
```

The async forms read from standard output/error using async methods on streams, so do not block the calling thread. The synchronous forms using JTF to keep things moving on the thread while the async operation completes.

This all comes together to code that resembles:

```c#
var output = _gitExecutable.GetOutput($"rev-list {parentHash} ^{childHash} --count");

var output = await _gitExecutable.GetOutputAsync(
    $"show {objectId}", cache: GitCommandCache, outputEncoding: LosslessEncoding);
```

The second example shows a change to command caching. Instead of having a global static cache, it's now an instance that may be passed into methods when results should be cached. This helps testing and reduces the number of method overloads.

A perk of this approach is that we can inject a `MockExecutable` into `GitModule` for testing. `GitModuleTest` has a bunch of new tests that look like this:

```c#
[Test]
public void GetParents_calls_correct_command_and_parses_response()
{
    using (_executable.StageOutput(
        $"log -n 1 --format=format:%P \"{Sha1}\"", // expected command line
        $"{Sha2} {Sha3}"))                         // staged process output
    {
        // verify method parsed response correctly
        Assert.AreEqual(
            new[] { Sha2, Sha3 },
            _gitModule.GetParents(Sha1));
    }
}
```

Such tests allow argument formulation and response parsing to exist in one place and be tested together. This will contribute significantly to the reduction in complexity of `GitModule` and related types such as `GitCommandHelpers`.

The new type `ArgumentString` acts as a union type over `string` and `ArgumentBuilder`. It can be extended to include future types such as the proposed `GitArgumentBuilder` if added. It reduces the number of overloads required to keep the API simple. It also communicates the intent of its values more precisely than `string`.

## Performance

Using `async`/`await` and JTF rather than `SynchronizedProcessReader` to retrieve data from processes results in many fewer threads servicing the application, and therefore less memory usage, GC work and a general increase in responsivity.

### Before timeline

Here's a timeline of what happens when we click one revision in the grid, then a few moments later, click another. We see a waterfall of threads being created to read various data from git processes, running for a while, then exiting.

image

### After timeline

The same activity produces a different pattern when using async IO:

image

Using the application for several minutes results in around 40 threads being created in total through the entire lifetime of the process with this change, versus several hundred beforehand.

Manually running a scripted set of user actions through the application for about a minute, before and after this PR, showed other quantitative improvements in profiling:

- Approx 10% reduction in memory traffic
- Approx 40% reduction in GC time

## Command log

The command log now includes _all_ processes launched by Git Extensions. We also see the exit code.
This gives a truer picture of what GE is actually doing behind the scenes.

## What this all enables

Opens the door for:

- Easier testing of code that launches processes, including all of `GitModule`
- Easier implementation of per-git-command configuration as discussed in https://github.com/gitextensions/gitextensions/pull/5182#issuecomment-409705693
- Simple asyncification of process launching code, including all of `GitModule` (just append `Async`)

## Miscellaneous

Changes made in service of the above:

- More unit tests
- `CommandCache` reuses `MruCache<TKey,TValue>`, simplifying its internals
- Some types moved to `GitExtUtils` for use between different assemblies
- Command log includes process exit code
- `ConsoleEmulatorOutputControl` and `EditboxBasedConsoleOutputControl` now log to the command log
- Plink related code moved to new `Plink`/`PlinkTests` classes
- `GitCommandHelpers` no longer has any methods for launching processes
- `GitModule` no longer has any methods for running arbitrary processes (callers use `Executable` and friends directly)
- `GitModule.VersionInUse` moved to `GitVersion.Current` as it had no dependency on `GitModule` and is useful elsewhere
- Added `GitModule.AddInteractiveAsync`/`ResetInteractiveAsync`
- Removed unused code

---

These changes have evolved slowly over the last month, and I feel pretty good about them now. I'll transition back to looking at regressions next week.

### CODE REVIEW

File: GitCommands/Git/GitModule.cs [hunk 1]
```diff
@@ -1644,15 +1478,6 @@ public string ResetFile(string file)
             });
         }
 
-        /// <summary>
-        /// Determines whether the given repository has index.lock file.
-        /// </summary>
-        /// <returns><see langword="true"/> is index is locked; otherwise <see langword="false"/>.</returns>
-        public bool IsIndexLocked()
```
Comment (RussKie) [inline GitCommands/Git/GitModule.cs]: what happened to this one? I can't seem to find the call site....

**Comment by drewnoakes** [comment]
@RussKie cheers for the review. I'll go through your feedback and rebase to fix conflicts soon. End of the day here now.

File: GitCommands/Git/Executable.cs [hunk 2]
```diff
@@ -0,0 +1,594 @@
...
+            _workingDir = workingDir;
+            _fileNameProvider = fileNameProvider;
+        }
+
+        /// <inheritdoc />
+        [PermissionSet(SecurityAction.Demand, Name = "FullTrust")]
+        public IProcess Start(ArgumentString arguments = default, bool createWindow = false, bool redirectInput = false, bool redirectOutput = false, Encoding outputEncoding = null)
+        {
+            // TODO should we set these on the child process only?
+            EnvironmentConfiguration.SetEnvironmentVariables();
```
[the remark this answers is not in the document]
Comment (RussKie) [reply GitCommands/Git/Executable.cs]: /cc: @jbialobr @KindDragon

File: GitUI/FormRemoteProcess.cs [hunk 3]
```diff
@@ -145,48 +144,37 @@ public static bool AskForCacheHostkey(IWin32Window owner, GitModule module, stri
...
+                return new Plink().Connect(remoteUrl);
             }
 
             return false;
         }
 
         protected override void DataReceived(object sender, TextEventArgs e)
         {
-            if (Plink)
+            if (Plink && e.Text.Contains("If you trust this host, enter \"y\" to add the key to"))
```
[the remark this answers is not in the document]
Comment (RussKie) [reply GitUI/FormRemoteProcess.cs]: Yes.
`GetOutputString()` gets called 3 times

File: GitUI/FormRemoteProcess.cs [hunk 4]
```diff
@@ -145,48 +144,37 @@ public static bool AskForCacheHostkey(IWin32Window owner, GitModule module, stri
...
             {
-                if (e.Text.Contains("If you trust this host, enter \"y\" to add the key to"))
+                if (MessageBox.Show(this, _fingerprintNotRegistredText.Text, _fingerprintNotRegistredTextCaption.Text, MessageBoxButtons.YesNo) == DialogResult.Yes)
```
[the remark this answers is not in the document]
Comment (RussKie) [reply GitUI/FormRemoteProcess.cs]: Alright, let's leave it as is

File: GitCommands/Git/Executable.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (drewnoakes) [reply GitCommands/Git/Executable.cs]: There was some previous discussion about environment variables in #4664 though nothing conclusive. It seems pretty harmless as is.

File: GitUI/FormRemoteProcess.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (drewnoakes) [reply GitUI/FormRemoteProcess.cs]: > GetOutputString() gets called 3 times

Fixed.

File: GitCommands/Git/GitModule.cs [hunk 5]
```diff
@@ -199,6 +201,31 @@ bool HasGitModulesFile(string path)
...
+////
+////        public string SubmodulePath
+////        {
+////            get
+////            {
+////                InitSuperproject();
+////                return _submodulePath;
+////            }
+////        }
+////>>>>>>> Add and use IExecutable and IProcess abstractions
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply GitCommands/Git/GitModule.cs]: Fixed. Junk from a merge.

**Comment by drewnoakes** [comment]
> I have been using this branch for some limited work for a couple of days. Briefly looked at the code. No real comment, looks good.

Thanks for testing it out @gerhardol.

> The only I have noticed is that Git Command Log occasionally has git calls without any arguments.
> I got this when opening Settings and just press OK.

I also see a parameterless `git` invocation when opening settings.

In fact I see four different process launches, all on the UI thread, which may contribute to slowness opening settings:

image

Exit code 1 is consistent with command line behaviour:

```
$ git
usage: ...
$ echo $?
1
```

I will look into this a bit more deeply today. I would like to add a checkbox to the command log form that enables capturing of stack traces for process invocations, so we can see which code paths are launching processes. That would be useful in this case. It should be disabled by default though, for perf.

> gitk still cannot be started too, that is not new

Are you talking about the tools menu item? It works for me.

image

**Comment by drewnoakes** [comment]
I added a checkbox that allows capturing call stacks.

image

Here's the output for the call to git with no arguments. I'll look further later.

```
File name:  C:\Program Files\Git\bin\git.exe
Arguments:  
Process ID: 16928
Started at: 13/08/2018 15:05:20
UI Thread?: True
Duration:   73.64ms
Exit code:  1
Call stack: 
   at GitCommands.Logging.CommandLog.LogProcessStart(String fileName, String arguments)
   at GitCommands.Executable.ProcessWrapper..ctor(String fileName, String arguments, String workDir, Boolean createWindow, Boolean redirectInput, Boolean redirectOutput, Encoding outputEncoding)
   at GitCommands.Executable.Start(ArgumentString arguments, Boolean createWindow, Boolean redirectInput, Boolean redirectOutput, Encoding outputEncoding)
   at GitCommands.ExecutableExtensions.<GetOutputAsync>d__3.MoveNext()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.Start[TStateMachine](TStateMachine& stateMachine)
   at GitCommands.ExecutableExtensions.GetOutputAsync(IExecutable executable, ArgumentString arguments, Byte[] input, Encoding outputEncoding, CommandCache cache, Boolean stripAnsiEscapeCodes)
   at GitCommands.ExecutableExtensions.<>c__DisplayClass2_0.<GetOutput>b__0()
   at Microsoft.VisualStudio.Threading.JoinableTaskFactory.ExecuteJob[T](Func`1 asyncMethod, JoinableTask job)
   at Microsoft.VisualStudio.Threading.JoinableTaskFactory.RunAsync[T](Func`1 asyncMethod, Boolean synchronouslyBlocking, JoinableTaskCreationOptions creationOptions)
   at Microsoft.VisualStudio.Threading.JoinableTaskFactory.Run[T](Func`1 asyncMethod, JoinableTaskCreationOptions creationOptions)
   at GitCommands.ExecutableExtensions.GetOutput(IExecutable executable, ArgumentString arguments, Byte[] input, Encoding outputEncoding, CommandCache cache, Boolean stripAnsiEscapeCodes)
   at GitUI.CommandsDialogs.SettingsDialog.CheckSettingsLogic.<>c.<SolveGitCommand>b__12_0(String cmd)
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source)
   at GitUI.CommandsDialogs.SettingsDialog.CheckSettingsLogic.SolveGitCommand(String possibleNewPath)
   at GitUI.CommandsDialogs.SettingsDialog.Pages.GitSettingsPage.GitPath_TextChanged(Object sender, EventArgs e)
   at System.Windows.Forms.Control.OnTextChanged(EventArgs e)
   at System.Windows.Forms.Control.set_Text(String value)
   at System.Windows.Forms.TextBoxBase.set_Text(String value)
   at System.Windows.Forms.TextBox.set_Text(String value)
   at GitUI.CommandsDialogs.SettingsDialog.Pages.GitSettingsPage.SettingsToPage()
   at GitUI.CommandsDialogs.SettingsDialog.SettingsPageBase.LoadSettings()
   at GitUI.CommandsDialogs.FormSettings.LoadSettings()
   at GitUI.CommandsDialogs.FormSettings.FormSettings_Shown(Object sender, EventArgs e)
   at System.Windows.Forms.Form.OnShown(EventArgs e)
   at System.Windows.Forms.Control.InvokeMarshaledCallbackHelper(Object obj)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Windows.Forms.Control.InvokeMarshaledCallback(ThreadMethodEntry tme)
   at System.Windows.Forms.Control.InvokeMarshaledCallbacks()
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.Form.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
   at System.Windows.Forms.UnsafeNativeMethods.DispatchMessageW(MSG& msg)
   at System.Windows.Forms.UnsafeNativeMethods.DispatchMessageW(MSG& msg)
   at System.Windows.Forms.Application.ComponentManager.System.Windows.Forms.UnsafeNativeMethods.IMsoComponentManager.FPushMessageLoop(IntPtr dwComponentID, Int32 reason, Int32 pvLoopData)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoopInner(Int32 reason, ApplicationContext context)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoop(Int32 reason, ApplicationContext context)
   at System.Windows.Forms.Form.ShowDialog(IWin32Window owner)
   at GitUI.CommandsDialogs.FormSettings.<>c__DisplayClass13_1.<ShowSettingsDialog>b__0()
   at GitCommands.AppSettings.<>c__DisplayClass26_0.<UsingContainer>b__0()
   at GitCommands.Settings.SettingsContainer`2.<>c__DisplayClass7_0.<LockedAction>b__0()
   at GitCommands.SettingsCache.<>c__DisplayClass3_0.<LockedAction>b__0()
   at GitCommands.SettingsCache.LockedAction[T](Func`1 action)
   at GitCommands.SettingsCache.LockedAction(Action action)
   at GitCommands.Settings.SettingsContainer`2.LockedAction(Action action)
   at GitCommands.AppSettings.UsingContainer(RepoDistSettings settingsContainer, Action action)
   at GitUI.CommandsDialogs.FormSettings.ShowSettingsDialog(GitUICommands uiCommands, IWin32Window owner, SettingsPageReference initialPage)
   at GitUI.GitUICommands.<>c__DisplayClass107_0.<StartSettingsDialog>g__Action|0()
   at GitUI.GitUICommands.DoActionOnRepo(IWin32Window owner, Boolean requiresValidWorkingDir, Boolean changesRepo, EventHandler`1 preEvent, EventHandler`1 postEvent, Func`1 action)
   at GitUI.GitUICommands.StartSettingsDialog(IWin32Window owner, SettingsPageReference initialPage)
   at GitUI.CommandsDialogs.FormBrowse.OnShowSettingsClick(Object sender, EventArgs e)
   at GitUI.CommandsDialogs.FormBrowse.ExecuteCommand(Int32 cmd)
   at ResourceManager.GitExtensionsFormBase.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.ContainerControl.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.ContainerControl.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.ContainerControl.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.ContainerControl.ProcessCmdKey(Message& msg, Keys keyData)
   at ResourceManager.GitExtensionsControl.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.TextBoxBase.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.PreProcessMessage(Message& msg)
   at System.Windows.Forms.Control.PreProcessControlMessageInternal(Control target, Message& msg)
   at System.Windows.Forms.Application.ThreadContext.PreTranslateMessage(MSG& msg)
   at System.Windows.Forms.Application.ComponentManager.System.Windows.Forms.UnsafeNativeMethods.IMsoComponentManager.FPushMessageLoop(IntPtr dwComponentID, Int32 reason, Int32 pvLoopData)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoopInner(Int32 reason, ApplicationContext context)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoop(Int32 reason, ApplicationContext context)
   at GitUI.GitUICommands.StartBrowseDialog(IWin32Window owner, String filter, ObjectId selectedCommit, Boolean startWithDashboard)
   at GitUI.GitUICommands.RunBrowseCommand(IReadOnlyList`1 args)
   at GitUI.GitUICommands.RunCommandBasedOnArgument(IReadOnlyList`1 args, IReadOnlyDictionary`2 arguments)
   at GitUI.GitUICommands.RunCommand(IReadOnlyList`1 args)
   at GitExtensions.Program.RunApplication()
   at GitExtensions.Program.Main()
```

**Comment by RussKie** [comment]
I'm not sure this is the best place to post it, but there are never ending calls to `git log -z`:

image

@gerhardol's screenshot above shows the same.

Perhaps it is not so much a _never ending call_ but its exit isn't tracked properly....

**Comment by gerhardol** [comment]
My comments were obviously not new in this PR, it is this PR that makes it possible to find the existing bugs.
No further comments on the code, I do not expect to see any significant issues when using the changes. 

@RussKie Regarding "running"; this was discussed before and drew had some explanation.
I guess we can start discussing such items related to where they were found, but when they are clearly separate, they should be discussed separately.

**Comment by drewnoakes** [comment]
> but its exit isn't tracked properly....

Bingo. There's a race condition between the process's `Exited` event and the calling of `Process.Dispose`. I looked a bit more at this yesterday and there are a few possible solutions, but none of them felt perfect so I decided to mull it over a while before returning to it.

It doesn't apply to the most common methods (eg. `GetOutput`) as they wait for the process to exit. Usages of `RunGitCmdDetached`, `GetOutputLines` and the like can dispose the `IProcess` object before it raises its `Exited` event. This means we don't capture a duration. The `running` text is shown when the `Duration` property is `null`.

Fixes include:

1. Forcing `IProcess.Dispose` to synchronously wait for the process to exit. At best this introduces some delay, at worst it deadlocks.
2. Assuming that `IProcess.Dispose` means the process exited. We won't have an exit code (it's a nullable `int` partly for this reason). This may end up under-reporting the process's running time.
3. Tracking additional status that the process was disposed before it exited, and showing 'unknown' or similar in the log.
4. Require all callers to explicitly wait for the process to exit.

I implemented 3, but backed it out as it felt clunky. I'm leaning towards 2, mostly because the issue of under-reporting time is not a critical issue. The first option potentially deadlocks. The fourth option puts the onus on the user of the API, so this issue would re-surface in future.

Changes in this area must be made carefully. The current `IProcess` implementation has been carefully designed around the behaviour of `Process` and such race conditions. Another advantage of option 2 above is that it shouldn't modify any of that carefully constructed code.

> Git Command Log occasionally has git calls without any arguments

I looked into this and the behaviour predates this PR. The changes here just make it visible by virtue of capturing more (I believe all) process executions.

**Comment by drewnoakes** [comment]
Pushed a commit that prevents the appearance of endless 'running' in the Command Log UI.

**Comment by RussKie** [comment]
A slightly different angle, is `Dispose` really necessary? Is there a
reason why we can't wait for `process.Exit` event and dispose of all
objects (if necessary)?

**Comment by drewnoakes** [comment]
Not disposing can lead to dangling process instances unfortunately, particularly in the case of exceptions when parsing. If we redirect the output of a process, start reading it, parse a chunk of it and throw, then the process keeps writing into a buffer that we never read from and lingers indefinitely. There are probably other failure modes there too.

**Comment by gerhardol** [comment]
Updates seem good to me. (Very brief look at the code, not tested but the PR seem ready for me.)

**Comment by drewnoakes** [comment]
Because the process won't exit if the consuming code doesn't consume the
redirected output streams, or explicit close them, as may happen during
posting errors.

If the event was reliable it would be enough to subscribe to it without any
kind of registration in a central collection.

[via phone]

File: GitCommands/Logging/CommandLog.cs [hunk 6]
```diff
@@ -55,22 +67,53 @@ internal CommandLogEntry(string fileName, string arguments, DateTime startedAt,
...
 
-            return $"{StartedAt:HH:mm:ss.fff} {duration,7} {pid} {(IsOnMainThread ? "UI" : "  ")} {fileName} {Arguments}";
+        public string Detail
+        {
+            get
+            {
+                var s = new StringBuilder();
+
+                s.Append("File name:  ").AppendLine(FileName);
+                s.Append("Arguments:  ").AppendLine(Arguments);
```
Comment (gerhardol) [inline GitCommands/Logging/CommandLog.cs]: Hey! Now I have to type git manually, cannot copy the complete command in one go!

**Comment by RussKie** [comment]
Please keep gitlog window enhancements separate

**Comment by gerhardol** [comment]
> Please keep gitlog window enhancements separate

In this case I allow scope creep. This PR has exposed other problems we have commented on, that turned out to be separate after Drew implemented support to investigate that. No major point to cut 
f49b51b just because of.

File: GitCommands/Logging/CommandLog.cs (hunk 6 above)
Comment (drewnoakes) [reply GitCommands/Logging/CommandLog.cs]: Ahh. The intent here was to show the full path for the executable somewhere.

Would a context menu "Copy" from the list box, with a Ctrl+C shortcut, work?

Comment (gerhardol) [reply GitCommands/Logging/CommandLog.cs]: I can live with typing "git" (or copying the file) when needed. 
A generic context menu with Copy would be an improvement, now you have to try and see if Ctrl-C works . Not important but helps the user, it is not ideal to view the stacktrace in the scrollbox. (I just want to view the stacktrace in a text editor, I do not want more support here!)

**Comment by drewnoakes** [comment]
> Please keep gitlog window enhancements separate

As requested, I pulled that change from this PR and created drewnoakes/gitextensions#7. It's now the second PR queued up behind this being merged, the other being drewnoakes/gitextensions#6 which was co-authored with @vbjay.

**Comment by RussKie** [comment]
I'm happy for it to be merged.

**Pull request merged**